src/server/managers/AppExternalComponentManager.ts

```diff
@@ -1,79 +1,129 @@
+        this.providedExternalComponents = new Map<string, Map<string, IExternalComponent>>();
+        this.appTouchedExternalComponents = new Map<string, Map<string, IExternalComponent>>();
     }
     /**
-     * Add the external component to the appTouchedComponents.
+     * Get all registered components.
+     */
+    public getProvidedExternalComponents(): Map<string, Map<string, IExternalComponent>> {
+        return this.providedExternalComponents;
+    }
+    /**
+     * Get all external components that apps have registered
+     * before, including disabled apps' external components.
+     */
+    public getAppTouchedExternalComponents(): Map<string, Map<string, IExternalComponent>> {
+        return this.appTouchedExternalComponents;
+    }
+    /**
+     * Get all external components of an app by specifying the appId.
      *
      * @param appId the id of the app
-     * @param externalComponent the external component to register
-     * @param addToProvidedComponents whether also add the external
-     * component to the providedComponents, the value is false by default.
      */
-    public registerComponent(appId: string, externalComponent: IExternalComponent, addToProvidedComponents: boolean = false): void {
-        this.appTouchedComponents.set(appId, externalComponent);
-
-        if (addToProvidedComponents) {
-            this.providedComponents.set(appId, externalComponent);
+    public getExternalComponents(appId: string): Map<string, IExternalComponent> {
+        if (this.appTouchedExternalComponents.has(appId)) {
+            return this.appTouchedExternalComponents.get(appId);
         }
+
+        return null;
     }
     /**
-     * Remove the external component from the providedComponents by appId.
+     * Add an external component to the appTouchedExternalComponents.
      *
      * @param appId the id of the app
+     * @param externalComponent the external component need to be added
      */
-    public unregisterComponent(appId: string): void {
-        if (this.providedComponents.has(appId)) {
-            this.providedComponents.delete(appId);
+    public addExternalComponent(appId: string, externalComponent: IExternalComponent): void {
+        if (externalComponent.appId !== appId) {
+            throw new ExternalComponentNotMatchWithAppError();
+        }
+
+        if (!this.appTouchedExternalComponents.get(appId)) {
+            this.appTouchedExternalComponents.set(appId, new Map(Object.entries({ [externalComponent.name]: externalComponent})));
+        } else {
+            const appExternalComponents = this.appTouchedExternalComponents.get(appId);
+            const touchedExternalComponent = appExternalComponents.get(externalComponent.name);
+
+            if (
+                touchedExternalComponent
+                && touchedExternalComponent.name === externalComponent.name
+                && JSON.stringify(touchedExternalComponent) !== JSON.stringify(externalComponent)
```

Comment:
I'm not sure if I understand the need for this, can you explain? If we are making the `name` part the unique aspect, then why would we care about if they are different on the internals? I don't think we should care about it as with this check it makes this call limit what can be done. What if I want the same component name but want to change the content when they make a setting change and only need/want that one component? What I just said could also be an undesired usage we want, but I think it would be fine 🤔

Reply:
As you mentioned in another [comment](https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/156#discussion_r332281537), **overwritten** if there's an external component with the same name instead of throwing an Error. This seems to be a better solution. I will change to it.

Reply:
Resolved in https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/150/commits/c250945c0052215f62835d6c07400c59a556265e and https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/150/commits/81d334f86757c6c76b42bebe9028f628061a7fc0